src/common/services/session/session.service.js

```diff
@@ -110,28 +116,50 @@ function session( $cookies, $rootScope, $http, envService ) {
 
   /* Private Methods */
   function updateCurrentSession( encoded_value ) {
-    var cortexSession = {};
+    let cortexSession = {};
     if ( angular.isDefined( encoded_value ) ) {
       cortexSession = jwtDecode( encoded_value );
     }
+    // Set expiration timeout
+    setSessionTimeout();
     // Copy new session into current session object
     angular.copy( cortexSession, session );
     // Update sessionSubject with new value
     sessionSubject.next( session );
   }
 
+  function setSessionTimeout() {
+    let encodedSession = $cookies.get( 'give-session' );
+    // Cancel current session timeout
+    if ( angular.isDefined( sessionTimeout ) ) {
+      $timeout.cancel( sessionTimeout );
+      sessionTimeout = undefined;
+    }
+    // Decode give-session cookie and set timeout based on expiration
+    if ( angular.isDefined( encodedSession ) ) {
+      let giveSession = jwtDecode( encodedSession ),
+        timeout = new Date( giveSession.exp * 1000 ) - Date.now();
+      if ( timeout > 0 ) {
+        sessionTimeout = $timeout( timeout );
+        sessionTimeout.then( () => {
+          updateCurrentSession( $cookies.get( 'cortex-session' ) );
```

Comment:
Browsers delete the expired cookies, so $cookies.get() returns undefined. getRole() uses this to downgrade the role. The cortex-session role will still say 'REGISTERED', but getRole() will return to correct role.
I have tested it (allowed a session to actually expire while on the page) and it worked correctly.